mitodl/micromasters
# Pearson response processing

_**Before starting review on this, please message me so I can share the documentation with you**_

---

#### What are the relevant tickets?

Fixes #1797 and #2422

#### What's this PR do?

This PR primarily adds processing of two types of Pearson result files. Each row in these files is a result of a corresponding row in a file we uploaded. The results are asynchronous and since Pearson only processes our uploads every 12 hours it can be a while between upload and results. You can think of this as RPC-over-SFTP. Pearson places these files on their SFTP server so we have to periodically poll and check for new ones. They also need to be removed once we've processed them.

The two types of files we're immediately interested in:

- VCDC - results to CDD (profile data) files we upload. Corresspond to `ExamProfile` records.
- EAD - results to EAD (exam authorization) files we upload. Correspond to `ExamAuthorization` records.

Both of these records go through a status lifecycle roughly `PENDING` -> `IN_PROGRESS` -> (`SUCCESS` | `FAILURE`). This code handles processing response files from Pearson to determine if our original request was a success or failure.

A slight complication is that these result files are contained in ZIP files and there's no documented convention of which result files show up in which zip files. The approach I've taken here is to basically walk all the zip files and then walk their contents file-by-file, handling the files that we currently support parsing for. ZIP files we don't consume fully are left on the server (in case we change our mind down the road and want to re-parse all historic data of that type). In my observations, the VCDC and EAD files always come together in the same zip file with no other file types.

#### Where should the reviewer start?

The meat of this implementation is in two places:

- `exams/pearson/readers.py` - implementation of VCDC- and EAC-specific readers. These parse TSV files into a list of rows, in this case using `namedtuple`s (could be classes, but that's overkill).
- `exams/pearson/download.py` - this code walks the SFTP server and downloads each zip file into a temporary directory. Each TSV file within is then extracted locally, parsed, and processed based on the filename prefix (e.g.` vcdc-2016-02-15-a.dat` is a VCDC file). Errors are logged and emailed.

#### How should this be manually tested?

To test first ensure you're enrolled in a course and have a fully filled out profile and then you can run this in shell (keep it open!):

```python
from django.contrib.auth.models import User
from exams.models import ExamProfile, ExamAuthorization
from courses.models import Course

user = User.objects.get(username=<YOU>)
course = Course.objects.get(id=<COURSE_ID>)

# fudge the records to sync
ep = ExamProfile.objects.create(profile=user.profile)
ea = ExamAuthorization.objects.create(user=user, course=course)

# sync each object type to the sftp server
tasks.export_exam_profiles()
tasks.export_exam_authorizations()
```

Then, run this command to simulate the processing of the files you just exported:

`docker-compose run web ./manage.py simulate_sftp_responses --poll`

That command will randomly assign success/failure response, write them out to a response file and put that file in a zip file on the local sftp container. This is basically a dumb simulation of what Pearson will do. You can check the sftp server to verify these files are in place :

```shell
sftp -P 2022 odl@localhost # password: 123
cd results
ls
```

Then you can finally run the real deal of this PR in the first shell you opened (you won't see much output since it all going to the logging module):

```python
tasks.batch_process_pearson_zip_files()
ea.refresh_from_db()
ep.refresh_from_db()
print((ea.id, ea.status))
print((ep.id, ep.status))
```

That should process and remove the zip files from the SFTP server and the respective record statuses should now be either success or failure matching what `simulate_sftp_responses` output (our model constants differ from the response file constants).

#### What GIF best describes this PR or how it makes you feel?

![Processing](http://i.giphy.com/5rVv3TyEzGcwM.gif)
